### DISCUSSION TITLE
Feature Request: Paragraph Numbering

### DISCUSSION THREAD

**Discussion by bhattmaulik** [quarto-dev/quarto-cli#1608] [Feature Requests]
In Quarto, we already have numbering in Table of Contents and Sections. I would also like to have paragraph numbering feature. This is especially useful in the legal field. For example, something is mentioned in Paragraph 1.2.3 [where 1 is the chapter number, 2 is the section number within the chapter (which would be denoted by Heading 2 in the Visual Editor), and the 3 is the number within this section of the chapter], which I can refer to at a later stage either in the same .qmd document, or in another .qmd document within the same pdf book.

This feature will also be useful for those writing thesis, where they have to give cross-reference from one chapter to another chapter. I understand that there is a feature of cross-referencing in Quarto. But in that feature, cross-referencing can be done only for those items, for which the writer adds an "index" entry. I am suggesting that every paragraph should have an index number, if some variable (e.g. paragraph-numbering) is set to TRUE.

How can it be implemented?

**Comment by mcanouil** [comment]
You could use low level headers and use custom CSS to make it look like paragraph.

```
---
title: "Untitled"
format: html
number-sections: true
include-before-body: 
  - text: |
      <style>
      h4 {
        margin-top: 0;
        margin-bottom: 1rem;
        display: block;
        margin-block-start: 1em;
        margin-block-end: 1em;
        margin-inline-start: 0px;
        margin-inline-end: 0px;
        font-weight: normal;
        font-size: var(--bs-body-font-size);
        font-weight: var(--bs-body-font-weight);
        line-height: var(--bs-body-line-height);
      }
      </style>
---

#### Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc finibus varius venenatis. Vivamus tincidunt, ex vitae sollicitudin ultrices, lectus massa sagittis eros, id interdum erat augue eget lacus. Vivamus nec nibh pulvinar, volutpat odio a, auctor arcu. Donec ac rutrum leo. Aliquam in diam pulvinar, aliquet mauris quis, laoreet lorem. Duis ut feugiat augue, sit amet tincidunt erat. Integer vitae arcu tristique, dapibus nibh at, rhoncus urna. Proin tincidunt massa eu ligula scelerisque feugiat. Vivamus et maximus tortor. Proin suscipit interdum erat in bibendum. Nulla facilisi. Duis commodo sapien aliquam erat pulvinar, a semper turpis porttitor. Nulla facilisi.

#### Vivamus pretium auctor luctus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc sodales sit amet quam a gravida. Praesent vestibulum felis at arcu blandit, a imperdiet turpis sodales. Sed elementum, ligula sit amet efficitur tempus, mi risus dictum erat, vel sollicitudin magna dolor et metus. Fusce volutpat orci lectus, a porttitor nisl semper eu. Fusce viverra enim et tellus tristique, sit amet ultrices odio efficitur. Mauris laoreet gravida nisi nec faucibus.

#### Sed elit nulla, lobortis et metus quis, porttitor venenatis ligula. Vestibulum laoreet feugiat lacus in lobortis. Nam magna magna, laoreet in massa a, egestas porttitor risus. Morbi dolor nulla, auctor eget dolor ut, porta efficitur arcu. Sed a aliquam diam, eget malesuada nisi. Aenean facilisis, magna ac cursus pretium, massa mauris fermentum leo, id gravida lorem leo nec massa. Vestibulum ultricies ac nunc et eleifend. Ut faucibus tortor condimentum dolor vehicula porta ac sed ligula. In ornare purus non est imperdiet vulputate. Morbi et magna ut augue lobortis gravida. Praesent aliquam vulputate massa, ut dignissim mi convallis sed. Vestibulum in laoreet augue, quis dapibus tortor. Mauris velit odio, condimentum ut metus nec, consectetur mattis elit. Phasellus ac lacus accumsan, mollis eros vel, efficitur urna.

#### Mauris sit amet lacus id metus tempus venenatis. Pellentesque eget ullamcorper arcu. Ut at nibh turpis. Cras non quam a lectus aliquet tincidunt. Cras quis mi ligula. Sed luctus risus nibh, non tempus lacus sollicitudin ac. Aenean auctor feugiat ligula eget luctus. Suspendisse ac ipsum ut magna finibus condimentum vitae id odio. Fusce lobortis, dolor vitae aliquet consequat, eros leo ullamcorper diam, quis laoreet libero ipsum at turpis.
```

Note: the way I included the modified css is not the recommended way, i.e., you should use a css file and use the yaml option `css: styles.css`

**Reply by bhattmaulik** [reply]
Thank you for this idea. It is working quite well in the HTML output. Now let me think of how to implement this in latex context, because my reports will anyway be generated in latex/pdf.

**Reply by mcanouil** [reply]
For LaTeX, there are plenty of resources online.
For example, you can redefine the paragraph command to include a counter, see https://latex.org/forum/viewtopic.php?t=34451.
You can find other example online.

**Reply by bhattmaulik** [reply]
This is a good starting point. I can work from there. The only difference is that this article has paragraph numbers right from the beginning. That means, if, for example, the first section has 10 paragraphs, these paragraphs will be numbered from 1 to 10. And when the second section starts, its first paragraph will be numbered 11. I want to restart the paragraph counter from each section. That way, paragraph numbers in my latex documents would have numbers like 1.1 to 1.10 and then 2.1 onward. I think I should be able to figure it out somehow.

**Reply by mcanouil** [reply]
Yes, you have to define the `\setcounter` as you need.
After a new quick search, this might be of interest: https://www.dickimaw-books.com/latex/admin/html/legaldocs.shtml or this https://9to5science.com/numbering-paragraphs-in-latex

**Comment by zackbatist** [comment]
I realize this is a few years old, but I want to second this feature request for native paragraph numbering. Paragraph numbering is essential for making high resolution references to specific parts of a document, and are especially important in html-based documents that do not feature page numbers.

@mcanouil's suggestion is ultimately a workaround, and requires substantial modifications to the document to gain this functionality.

**Reply by mcanouil** [reply]
Please use post reactions and upvotes on the main post.
This is better than commenting and pinging.

**Reply by zackbatist** [reply]
I apologize if this was interpreted as spamming. However I intended to add an additional argument for why this is important.

**Reply by mcanouil** [reply]
No need to apologise. Nothing wrong was done.
I'm just saying that upvoting discussions (the arrow) and adding thumbs up on issues is the best way to show interest in something.
Developers/maintainers can filter/sort based on that.
A comment triggers notifications to everyone that subscribed and does not really increase visibility.
Note that the above is true everywhere on GitHub.
